### PR TITLE
Tarjan

### PR DESCRIPTION
**Pull request by grant-d**
Formatting
Comments
Use explicit IEqualityComparer

Please read all comments - several questions

### CODE REVIEW

File: src/SourceCode.Clay.Algorithms/Edge.cs [hunk 1]
```diff
@@ -12,40 +12,51 @@
...
     public readonly struct Edge<T> : IEquatable<Edge<T>>, IStructuralEquatable
     {
+        private readonly IEqualityComparer<T> _equalityComparer;
+
         public T From { get; }
+
         public T To { get; }
 
-        public Edge(T from, T to)
+        public Edge(T from, T to, IEqualityComparer<T> equalityComparer)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Edge.cs]: Not sure if we prefer separate overloads or default params

File: src/SourceCode.Clay.Algorithms/Edge.cs [hunk 2]
```diff
@@ -12,40 +12,51 @@
...
+        public Edge(T from, T to)
+            : this(from, to, null)
+        { }
 
         public override string ToString() => $"{From} -> {To}";
 
         public override bool Equals(object obj)
-            => obj is Edge<T> o && Equals(o);
+            => obj is Edge<T> other
+            && Equals(other);
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Edge.cs]: Consistent with formatting in rest of Clay

File: src/SourceCode.Clay.Algorithms/Edge.cs [hunk 3]
```diff
@@ -12,40 +12,51 @@
...
 
         public bool Equals(Edge<T> other)
-            => EqualityComparer<T>.Default.Equals(From, other.From)
-            && EqualityComparer<T>.Default.Equals(To, other.To);
+            => _equalityComparer.Equals(From, other.From)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Edge.cs]: Explicit comparer is more flexible than assuming .Default

File: src/SourceCode.Clay.Algorithms/Edge.cs [hunk 4]
```diff
@@ -12,40 +12,51 @@
...
         }
 
-        public int GetHashCode(IEqualityComparer comparer)
+        public static bool operator ==(Edge<T> edge1, Edge<T> edge2) 
+            => edge1.Equals(edge2);
+
+        public static bool operator !=(Edge<T> edge1, Edge<T> edge2) 
+            => !(edge1 == edge2);
+
+        int IStructuralEquatable.GetHashCode(IEqualityComparer comparer)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Edge.cs]: Do we actually need IStructuralEquatable

File: src/SourceCode.Clay.Algorithms/Graph.Nested.cs [hunk 5]
```diff
@@ -12,26 +12,26 @@
 
 namespace SourceCode.Clay.Algorithms
 {
-    public partial struct Graph<T>
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.Nested.cs]: Redundant decoration

File: src/SourceCode.Clay.Algorithms/Graph.cs [hunk 6]
```diff
@@ -39,6 +37,8 @@ public Graph(IEqualityComparer<T> equalityComparer)
 
         public void Add(T from, T to)
         {
+            var nodes = _nodes;
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.cs]: Capture

File: src/SourceCode.Clay.Algorithms/TreeNode.cs [hunk 7]
```diff
@@ -18,57 +18,66 @@ namespace SourceCode.Clay.Algorithms
...
+            : this(node, equalityComparer, (IReadOnlyList<int>)hierarchy)
+        { }
 
-        public static bool operator ==(TreeNode<T> node1, TreeNode<T> node2) => node1.Equals(node2);
+        public TreeNode(T node, IReadOnlyList<int> hierarchy)
+            : this(node, null, hierarchy)
+        { }
 
-        public static bool operator !=(TreeNode<T> node1, TreeNode<T> node2) => !(node1 == node2);
+        public TreeNode(T node, params int[] hierarchy)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/TreeNode.cs]: Can't use .Default for IEqualityComparer, so went with overloads

File: src/SourceCode.Clay.Algorithms/TreeNode.cs [hunk 8]
```diff
@@ -18,57 +18,66 @@ namespace SourceCode.Clay.Algorithms
...
         {
             if (Hierarchy is null) return other.Hierarchy is null;
             if (other.Hierarchy is null) return false;
 
-            if (!EqualityComparer<T>.Default.Equals(Node, other.Node)) return false;
             if (Hierarchy.Count != other.Hierarchy.Count) return false;
+            if (!_equalityComparer.Equals(Node, other.Node)) return false;
 
-            for (var i = Hierarchy.Count - 1; i >= 0; i--)
+            for (var i = 0; i < Hierarchy.Count; i++)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/TreeNode.cs]: Not sure why it was reversed. Not idiomatic but you may have had a reason

File: src/SourceCode.Clay.Algorithms/Graph.cs [hunk 9]
```diff
@@ -28,13 +29,18 @@ public static class Graph
         private readonly ConcurrentDictionary<T, Node> _nodes;
         private readonly IEqualityComparer<T> _equalityComparer;
 
-        public Graph(IEqualityComparer<T> equalityComparer)
+        public Graph(int capacity, IEqualityComparer<T> equalityComparer)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.cs]: Perf/scale: Explicit capacity

File: src/SourceCode.Clay.Algorithms/Graph.cs [hunk 10]
```diff
@@ -28,13 +29,18 @@ public static class Graph
...
-            _equalityComparer = equalityComparer;
-            _nodes = new ConcurrentDictionary<T, Node>(1, 1000, equalityComparer);
+            if (capacity <= 0) throw new ArgumentOutOfRangeException(nameof(capacity));
+
+            _equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
+            _nodes = new ConcurrentDictionary<T, Node>(1, capacity, _equalityComparer);
         }
 
+        public Graph(IEqualityComparer<T> equalityComparer)
+            : this(1000, equalityComparer)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.cs]: 1000 is arbitrary?

File: src/SourceCode.Clay.Algorithms/Edge.cs [hunk 11]
```diff
@@ -12,40 +12,47 @@
...
     public readonly struct Edge<T> : IEquatable<Edge<T>>, IStructuralEquatable
     {
+        private readonly IEqualityComparer<T> _equalityComparer;
+
         public T From { get; }
+
         public T To { get; }
 
-        public Edge(T from, T to)
+        public Edge(T from, T to, IEqualityComparer<T> equalityComparer = null)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Edge.cs]: Explicit comparer is more flexible

File: src/SourceCode.Clay.Algorithms/Graph.Nested.cs [hunk 12]
```diff
@@ -12,26 +12,26 @@
...
+    partial struct Graph<T> // .Nested
     {
         [Flags]
         private enum EdgeOptions : byte
         {
-            None = 0b0000_0000,
-            FindExitsExecuted = 0b0001_0000
+            None = 0b_0000_0000,
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.Nested.cs]: Easier on the eye to parse

File: src/SourceCode.Clay.Algorithms/Graph.cs [hunk 13]
```diff
@@ -5,40 +5,41 @@
...
     {
         private readonly ConcurrentDictionary<T, Node> _nodes;
         private readonly IEqualityComparer<T> _equalityComparer;
 
-        private static Node CreateValue(T key) => new Node();
-
-        private Node GetOrAdd(T key) => _nodes.GetOrAdd(key, CreateValue);
-
-        public Graph(IEqualityComparer<T> equalityComparer)
+        public Graph(int capacity, IEqualityComparer<T> equalityComparer)
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.cs]: Perf/scale: Explicit capacity. 1000 is arbitrary

File: src/SourceCode.Clay.Algorithms/Graph.cs [hunk 14]
```diff
@@ -5,40 +5,41 @@
...
         {
-            if (equalityComparer == null) equalityComparer = EqualityComparer<T>.Default;
-            _equalityComparer = equalityComparer;
-            _nodes = new ConcurrentDictionary<T, Node>(1, 1000, equalityComparer);
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.cs]: 1000 is arbitrary

File: src/SourceCode.Clay.Algorithms/Graph.cs [hunk 15]
```diff
@@ -48,6 +49,10 @@ public void Add(T from, T to)
 
             fromState.Edges.TryAdd(to, EdgeOptions.None);
             _nodes[to] = GetOrAdd(to).SetOptions(add: NodeOptions.Descendant);
+
+            Node GetOrAdd(T key) => nodes.GetOrAdd(key, CreateValue);
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.cs]: Local functions

File: src/SourceCode.Clay.Algorithms/Graph.cs [hunk 16]
```diff
@@ -48,6 +49,10 @@ public void Add(T from, T to)
...
+
+            Node CreateValue(T key) => new Node();
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/Graph.cs]: Not sure we actually need this?

File: src/SourceCode.Clay.Algorithms/TreeNode.cs [hunk 17]
```diff
@@ -18,57 +18,66 @@ namespace SourceCode.Clay.Algorithms
...
 
-        public TreeNode<T> Clone()
-        {
-            if (Hierarchy is null) return default;
-            return new TreeNode<T>(Node, Hierarchy.ToArray());
-        }
+        public TreeNode<T> Clone() 
+            => Hierarchy is null ? 
+            default : 
+            new TreeNode<T>(Node, _equalityComparer, Hierarchy.ToArray()); // Clone via ToArray
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/TreeNode.cs]: It wasn't obvious why `ToArray` was needed

File: src/SourceCode.Clay.Algorithms/TreeNode.cs [hunk 18]
```diff
@@ -80,7 +89,7 @@ public override int GetHashCode()
         {
             var hashCode = -1781160927;
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms/TreeNode.cs]: Pity we can't use your new stuff. Does this have to be `NetStandard`?

File: src/SourceCode.Clay.Algorithms.Tests/GraphTests.cs [hunk 19]
```diff
@@ -66,7 +66,7 @@ public void Graph_ToRepresentativeForest_Scenario1()
                       \--/
             */
 
-            var g = Graph.Create<int>();
+            var g = Graph.Create<int>(12);
```
Comment (grant-d) [inline src/SourceCode.Clay.Algorithms.Tests/GraphTests.cs]: See related change in factory method

File: src/SourceCode.Clay.Algorithms/Graph.cs (hunk 16 above)
Comment (jcdickinson) [reply src/SourceCode.Clay.Algorithms/Graph.cs]: This prevents an allocation.

File: src/SourceCode.Clay.Algorithms/TreeNode.cs (hunk 17 above)
Comment (jcdickinson) [reply src/SourceCode.Clay.Algorithms/TreeNode.cs]: Hierarchy changes across `onTreeNode` invocations (the same `List<TreeNode>` is used for the entire execution).

File: src/SourceCode.Clay.Algorithms/TreeNode.cs (hunk 8 above)
Comment (jcdickinson) [reply src/SourceCode.Clay.Algorithms/TreeNode.cs]: Hierarchies have a greater probability of being different starting from the end. Please revert.

File: src/SourceCode.Clay.Algorithms/TreeNode.cs [hunk 20]
```diff
@@ -80,7 +89,7 @@ public override int GetHashCode()
...
 
-            hashCode = hashCode * -1521134295 + EqualityComparer<T>.Default.GetHashCode(Node);
+            hashCode = hashCode * -1521134295 + _equalityComparer.GetHashCode(Node);
```
Comment (jcdickinson) [inline src/SourceCode.Clay.Algorithms/TreeNode.cs]: `_equalityComparer` may be null (struct). Introduce a local `var equalityComparer = _equalityComparer ?? EqualityComparer<T>.Default`.

File: src/SourceCode.Clay.Algorithms/Edge.cs (hunk 3 above)
Comment (jcdickinson) [reply src/SourceCode.Clay.Algorithms/Edge.cs]: `var equalityComparer = _equalityComparer ?? EqualityComparer<T>.Default`.

File: src/SourceCode.Clay.Algorithms/Edge.cs [hunk 21]
```diff
@@ -12,40 +12,51 @@
...
 
         public bool Equals(Edge<T> other)
-            => EqualityComparer<T>.Default.Equals(From, other.From)
-            && EqualityComparer<T>.Default.Equals(To, other.To);
+            => _equalityComparer.Equals(From, other.From)
+            && _equalityComparer.Equals(To, other.To);
 
         public override int GetHashCode()
         {
             var hashCode = -1781160927;
```
Comment (jcdickinson) [inline src/SourceCode.Clay.Algorithms/Edge.cs]: `var equalityComparer = _equalityComparer ?? EqualityComparer<T>.Default`.

File: src/SourceCode.Clay.Algorithms/TreeNode.cs [hunk 22]
```diff
@@ -18,57 +18,66 @@ namespace SourceCode.Clay.Algorithms
...
+            => obj is TreeNode<T> other
+            && Equals(other);
 
         public bool Equals(TreeNode<T> other)
         {
             if (Hierarchy is null) return other.Hierarchy is null;
             if (other.Hierarchy is null) return false;
 
-            if (!EqualityComparer<T>.Default.Equals(Node, other.Node)) return false;
             if (Hierarchy.Count != other.Hierarchy.Count) return false;
```
Comment (jcdickinson) [inline src/SourceCode.Clay.Algorithms/TreeNode.cs]: `var equalityComparer = _equalityComparer ?? EqualityComparer<T>.Default`.

**Pull request merged**